lib/phoenix_live_view/async.ex

```diff
@@ -3,44 +3,47 @@ defmodule Phoenix.LiveView.Async do
 
   alias Phoenix.LiveView.{AsyncResult, Socket, Channel}
 
-  defp validate_function_env(func, op, env) do
-    warn =
-      # TODO: Remove conditional once we require Elixir v1.14+
-      if Version.match?(System.version(), ">= 1.14.0") do
-        fn msg, env -> IO.warn(msg, env) end
-      else
-        fn msg, _env -> IO.warn(msg) end
-      end
+  defp warn_socket_access(op, warn) do
+    warn.("""
+    you are accessing the LiveView Socket inside a function given to #{op}.
 
-    # prevent false positives, for example
-    # start_async(socket, :foo, function_that_returns_the_anonymous_function(socket))
-    if match?({:&, _, _}, func) or match?({:fn, _, _}, func) do
-      Macro.prewalk(func, fn
-        {:socket, meta, nil} ->
-          warn.(
-            """
-            you are accessing the LiveView Socket inside a function given to #{op}.
+    This is an expensive operation because the whole socket is copied to the new process.
 
-            This is an expensive operation because the whole socket is copied to the new process.
+    Instead of:
 
-            Instead of:
+        #{op}(socket, :key, fn ->
+          do_something(socket.assigns.my_assign)
+        end)
 
-                #{op}(socket, :key, fn ->
-                  do_something(socket.assigns.my_assign)
-                end)
+    You should do:
 
-            You should do:
+        my_assign = socket.assigns.my_assign
 
-                my_assign = socket.assigns.my_assign
+        #{op}(socket, :key, fn ->
+          do_something(my_assign)
+        end)
 
-                #{op}(socket, :key, fn ->
-                  do_something(my_assign)
-                end)
+    For more information, see https://hexdocs.pm/elixir/1.16.1/process-anti-patterns.html#sending-unnecessary-data.
+    """)
+  end
 
-            For more information, see https://hexdocs.pm/elixir/1.16.1/process-anti-patterns.html#sending-unnecessary-data.
-            """,
-            Keyword.take(meta, [:line, :column]) ++ [line: env.line, file: env.file]
-          )
+  defp validate_function_env(func, op, env) do
+    # prevent false positives, for example
+    # start_async(socket, :foo, function_that_returns_the_anonymous_function(socket))
+    if match?({:&, _, _}, func) or match?({:fn, _, _}, func) do
+      Macro.prewalk(func, fn
+        {:socket, meta, nil} ->
+          warn_socket_access(op, fn msg ->
+            # TODO: Remove conditional once we require Elixir v1.14+
+            meta =
+              if Version.match?(System.version(), ">= 1.14.0") do
+                Keyword.take(meta, [:line, :column]) ++ [line: env.line, file: env.file]
+              else
+                []
```

Comment:
```suggestion
                Macro.Env.stacktrace(env)
```

better than nothing :)